Feature/remove unnecessary string handling in logging

fredrikhaglund opened the pull request:
Please take a moment to fill out the following:

Fixes issue #1089.

Changes proposed in this pull request:
- Use EmptyLogger as default in Release mode
- Add #define DEBUG to DebugLogger classto make it work in nuget release

dnfclas commented:
@fredrikhaglund,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by .NET Foundation. We will now review your pull request.
_Thanks,
.NET Foundation Pull Request Bot_

brianlagunas commented on Source/Prism/Logging/DebugLogger.cs:
```diff
@@ -5,6 +5,7 @@
 
 namespace Prism.Logging
 {
+#if DEBUG
```
Remove this check.  This will break anyone using this class as Prism is not released in Debug mode, so this class will now be missing from the assembly.

fredrikhaglund commented:
You are aware that this class does not work at all unless compiled in debug mode? Debug.WriteLine is removed by the compiler when compiled in release mode. 

My whole point is that I used this class in my project and wasted time figuring out why nothing was logged. 

Another option that does not introduce a breaking change by removing the class would be to add an obsolete attribute with a warning when compiled in release mode. Better?

brianlagunas commented:
I've thought about this a little bit and probably the best approach would be to use `#define DEBUG` in the logger so it will work.  This means we could remove all the other `#if` checks

brianlagunas commented:
Let me think about this a little more. Maybe we should remove the DebugLogger. I did it to make my life easier for developing, but it provides no value to anyone using the class outside of debugging the source code.

dansiegel commented:
I think perhaps the better idea would be to add the DebugLogger to the project templates. This would solve a few problems for developers...

brianlagunas commented:
That's actually not a bad idea.

fredrikhaglund commented:
I added #define DEBUG and removed conditional compilation to DebugLogger class.

Removing the class or hiding it with a conditional compile would be a breaking change so this is better.

brianlagunas commented:
With the DEBUG defined, there should be no need to have the conditional EmptyLogger now right?

fredrikhaglund commented:
Changing so the default is always the EmptyLogger and updating unit test to use DebugLogger? 

Leaving the DebugLogger as default will cause unneeded string handling and debug logging in release builds of Apps. Not a big deal but everything that can help improve performance in mobile apps is welcome...

dansiegel commented:
As I recall `System.Diagnostics.Debug` is useless in unit tests. I believe the best thing is to remove the DebugLogger altogether. If we want output from the unit tests then we want to use `ITestOutputHelper`.

brianlagunas commented:
Thanks for the PR
